pkg/provisioner/ironic/factory.go

```diff
@@ -0,0 +1,219 @@
+package ironic
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/go-logr/logr"
+	"github.com/gophercloud/gophercloud"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	logz "sigs.k8s.io/controller-runtime/pkg/log/zap"
+
+	metal3v1alpha1 "github.com/metal3-io/baremetal-operator/apis/metal3.io/v1alpha1"
+	"github.com/metal3-io/baremetal-operator/pkg/bmc"
+	"github.com/metal3-io/baremetal-operator/pkg/provisioner"
+	"github.com/metal3-io/baremetal-operator/pkg/provisioner/ironic/clients"
+)
+
+type ironicProvisionerFactory struct {
+	log    logr.Logger
+	config ironicConfig
+
+	// Keep pointers to ironic and inspector clients configured with
+	// the global auth settings to reuse the connection between
+	// reconcilers.
+	clientIronic    *gophercloud.ServiceClient
+	clientInspector *gophercloud.ServiceClient
+}
+
+func NewProvisionerFactory() provisioner.Factory {
+	factory := ironicProvisionerFactory{}
+
+	factory.log = logz.New().WithName("provisioner").WithName("ironic")
+
+	err := factory.init()
+	if err != nil {
+		factory.log.Error(err, "Cannot start ironic provisioner")
+		os.Exit(1)
+	}
+	return factory
+}
+
+func (f *ironicProvisionerFactory) init() error {
+	ironicAuth, inspectorAuth, err := clients.LoadAuth()
+	if err != nil {
+		return err
+	}
+
+	f.config, err = loadConfigFromEnv()
+	if err != nil {
+		return err
+	}
+
+	ironicEndpoint, inspectorEndpoint, err := loadEndpointsFromEnv()
+	if err != nil {
+		return err
+	}
+
+	tlsConf := loadTLSConfigFromEnv()
+
+	f.log.Info("ironic settings",
+		"endpoint", ironicEndpoint,
+		"ironicAuthType", ironicAuth.Type,
+		"inspectorEndpoint", inspectorEndpoint,
+		"inspectorAuthType", inspectorAuth.Type,
+		"deployKernelURL", f.config.deployKernelURL,
+		"deployRamdiskURL", f.config.deployRamdiskURL,
+		"deployISOURL", f.config.deployISOURL,
+	)
+
+	f.clientIronic, err = clients.IronicClient(
+		ironicEndpoint, ironicAuth, tlsConf)
+	if err != nil {
+		return err
+	}
+
+	f.clientInspector, err = clients.InspectorClient(
+		inspectorEndpoint, inspectorAuth, tlsConf)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func newTestProvisionerFactory() ironicProvisionerFactory {
+	return ironicProvisionerFactory{
+		log: logf.Log,
+		config: ironicConfig{
+			deployKernelURL:  "http://deploy.test/ipa.kernel",
+			deployRamdiskURL: "http://deploy.test/ipa.initramfs",
+			deployISOURL:     "http://deploy.test/ipa.iso",
+			maxBusyHosts:     20,
+		},
+	}
+}
+
+// A private function to construct an ironicProvisioner (rather than a
+// Provisioner interface) in a consistent way for tests.
+func newProvisionerWithSettings(host metal3v1alpha1.BareMetalHost, bmcCreds bmc.Credentials, publisher provisioner.EventPublisher, ironicURL string, ironicAuthSettings clients.AuthConfig, inspectorURL string, inspectorAuthSettings clients.AuthConfig) (*ironicProvisioner, error) {
```

Comment:
If this function it's just in tests it could be moved into one of the `*_test.go` file under `pkg/provisioner/ironic`, for example `ironic_test.go`